sdk/core/Azure.Core.Experimental/src/Variant/Value.cs

```diff
@@ -0,0 +1,1347 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Azure {
+    /// <summary>
+    /// Used to store primitive values without boxing, and other intstances.
+    /// </summary>
+    public readonly partial struct Value
+    {
+        private readonly Union _union;
+        private readonly object? _object;
+
+        /// <summary>
+        /// Creates instance.
+        /// </summary>
+        /// <param name="value"></param>
+        public Value(object? value)
+        {
+            _object = value;
+            _union = default;
+        }
+
+        /// <summary>
+        /// Type of the instance stored in this value.
+        /// </summary>
+        public readonly Type? Type
+        {
+            get
+            {
+                Type? type;
+                if (_object is null)
+                {
+                    type = null;
+                }
+                else if (_object is TypeFlag typeFlag)
+                {
+                    type = typeFlag.Type;
+                }
+                else
+                {
+                    type = _object.GetType();
+
+                    if (_union.UInt64 != 0 && type.IsArray)
+                    {
+                        // We have an ArraySegment
+                        if (type == typeof(byte[]))
+                        {
+                            type = typeof(ArraySegment<byte>);
+                        }
+                        else if (type == typeof(char[]))
+                        {
+                            type = typeof(ArraySegment<char>);
+                        }
+                        else
+                        {
+                            ThrowInvalidOperation();
+                        }
+                    }
+                }
+
+                return type;
+            }
+        }
+
+        [DoesNotReturn]
+        private static void ThrowInvalidCast() => throw new InvalidCastException();
+
+        [DoesNotReturn]
+        private static void ThrowArgumentNull(string paramName) => throw new ArgumentNullException(paramName);
+
+        [DoesNotReturn]
+        private static void ThrowInvalidOperation() => throw new InvalidOperationException();
+
+        #region Byte
+        /// <summary>
+        /// Stores byte in this value.
+        /// </summary>
+        /// <param name="value"></param>
+        public Value(byte value)
+        {
+            this = default;
```

Comment:
Curious: Why do this?

Reply:
Without it the complier complains that all fields need to be assigned. The reason is that the bye filed is only 8 bits and then there are all the other bits that store for example long, int, etc. that would be not assigned.

Reply:
Add as a comment?